lua/cfc_time/server/time/sv_mysql.lua

```diff
@@ -0,0 +1,265 @@
+function storage:BuildSessionUpdate( data, id )
+    local updateSection = "UPDATE sessions "
+    local setSection = "SET "
+    local whereSection = string.format(
+        "WHERE id = %s AND realm = '%s'",
+        id, self.realm
+    )
+
+    -- TODO: Have a safeguard here for invalid keys?
+    local count = table.Count( data )
+    local idx = 1
+
+    -- FIXME: This will break because we pass strings/ints interchangeably, but strings need quotes around them
+    for k, v in pairs( data ) do
+        local newSet = k .. " = " .. v
+        
+        if idx ~= count then
+            -- Add a comma if it isn't the last one
+            newSet = newSet .. ","
+        else
+            -- Add a space if it's the last one
+            newSet = newSet .. " "
+        end
+        
+        setSection = setSection .. newSet
+        idx = idx + 1
+    end
+
+    local query = updateSection .. setSection .. whereSection
+
+    logger:info( "Created sessions update query!" )
+    logger:info( query )
+
+    return self.database:escape( query )
+end
+
+--[ API Begins Here ]--
+
+function storage:UpdateBatch( batchData )
+    if not batchData then return end
+    if table.Count( batchData ) == 0 then return end
+
+    local transaction = storage:InitTransaction()
+
+    for sessionId, data in pairs( batchData ) do
+        local updateStr = self:BuildSessionUpdate( data, sessionId )
+        local query = self.database:query( updateStr )
+
+        transaction:addQuery( query )
+    end
+
+    transaction:start()
+end
+
+function storage:GetTotalTime( steamId, callback )
+    local onSuccess = function( _, data )
+        callback( data )
+    end
+
+    local query = self:Prepare( "totalTime", onSuccess, steamId )
+
+    query:start()
+end
+
+function storage:CreateSession( callback, steamId, sessionStart, sessionEnd, duration )
+    local newSession = self:Prepare( "newSession", callback, steamId, sessionStart, sessionEnd, duration )
+    newSession:start()
+end
+
+-- Takes a steamid, a session start timestamp, and a callback, then:
+--  - Creates a new user (if needed)
+--  - Creates a new session with given values
+-- Calls callback with a structure containing:
+--  - sessionId (the id of the newly created session)
+--  - totalTime (the calculated total playtime)
+function storage:PlayerInit( steamId, sessionStart, callback )
+    logger:info( "Receiving PlayerInit call for: " .. tostring( steamId ) )
+    local transaction = storage:InitTransaction()
+
+    local newUser = self:Prepare( "newUser", nil, steamId )
+    local newSession = self:Prepare( "newSession", nil, steamId, sessionStart, nil, 0 )
+    local totalTime = self:Prepare( "totalTime", nil, steamId )
+    local sessionData = self:Prepare( "latestSession", nil, steamId )
+
+    transaction:addQuery( newUser )
+    transaction:addQuery( newSession )
+    transaction:addQuery( totalTime )
+    transaction:addQuery( sessionData )
+
+    transaction.onSuccess = function( q, data )
```

Comment:
```suggestion
    transaction.onSuccess = function( t, data )
```
First arg here is the transaction, not the query

Also, to get this bit of code to work, you need to unpack the query from the transaction using `transaction:getQueries()` then call `query:getData()` on the relevant queries
`getData` will return a list of results, even if there's only 1 result (which is the case for totalTime and sessionData), so you'll need to do something like
`transaction:getQueries()[3]:getData()[1]["SUM(duration)"]` or write some function to unpack every queries data in a tranction into one big table, as follows:

```lua
local function getTransactionData( t )
    local data = {}

    for k, query in pairs( t:getQueries() ) do
        data[k] = query:getData()    
    end

    return data
end
```
and
```lua
local data = getTransactionData( t )
print( data[3][1]["SUM(duration"] )
```